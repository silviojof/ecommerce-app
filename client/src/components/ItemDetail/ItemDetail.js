import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { fetchDetail, clearDetail } from '../../redux/ducks/productDetail';
import Breadcrumb from '../Breadcrumb';
import Loader from '../Loader';

import styles from './ItemDetail.module.scss';

const parseCondition = status => {
    let condition;
    switch(status) {
        case 'new':
        case 'nuevo':
            condition = 'Nuevo';
            break;
        case 'used':
            condition = 'Usado';
            break;
        default:
            condition = status;
    }
    return condition;
}

export const ItemDetail = ({ location, match, fetchDetails, details, clearDetail, isLoading }) => {
    useEffect(() => {
        if (match.params.id) {
            fetchDetails(match.params.id)
        }
        return () => {
            clearDetail();
        };
      }, [location.search, match.params.id]);

    if (isLoading) {
        return (
            <div data-testid="loader" className={styles.loader}>
                <Loader />
            </div>
        )
    }
    
    return (
            <div className={styles.container}>
                <Breadcrumb />
                {
                    details !== null &&
                    <section className={styles.detail} data-testid="content">
                        <img
                            className={styles.image}
                            src={details.picture} alt="product"
                        />
                        <div>
                            <span
                                className={styles.caption}
                            >
                                {`${parseCondition(details.condition)} - ${details.sold_quantity} vendidos`}
                            </span>
                            <h5 className={styles.title}>{details.title}</h5>
                            <h3 className={styles.value}>
                                {`${details.price.currency} ${details.price.amount.toLocaleString('es-AR')}`}
                                <span>{details.price.decimals || '00'}</span>
                            </h3>
                            <button className={styles.button}>Comprar</button>
                        </div>
                        <div className={styles.description}>
                            <h4>Descripci√≥n del produto</h4>
                            <p>{details.description}</p>
                        </div>
                    </section>
                }
            </div>
    )
};

ItemDetail.propTypes = {
    location: PropTypes.object.isRequired,
    match: PropTypes.object.isRequired,
    fetchDetails: PropTypes.func.isRequired,
    details: PropTypes.shape({
        picture: PropTypes.string.isRequired,
        condition: PropTypes.string.isRequired,
        sold_quantity: PropTypes.number.isRequired,
        price: PropTypes.shape({
            currency: PropTypes.string.isRequired,
            amount: PropTypes.number.isRequired,
            decimals: PropTypes.number.isRequired,
        }),
        description: PropTypes.string.isRequired,
    }),
    clearDetail: PropTypes.func.isRequired,
    isLoading: PropTypes.bool.isRequired,
};

ItemDetail.defaultProps = {
    details: {
        picture: '',
        condition: '',
        sold_quantity: 0,
        price: {
            currency: '',
            amount: 0,
            decimals: 0,
        },
        description: '',
    },
};

const mapStateToProps = (state) => {
    return {
        details: state.productDetail.details,
        isLoading: state.productDetail.isLoadingDetails
    }
}

const mapDispatchToProps = dispatch => {
    return {
      fetchDetails: args => dispatch(fetchDetail(args)),
      clearDetail: () => dispatch(clearDetail()),
    }
  }

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(ItemDetail);