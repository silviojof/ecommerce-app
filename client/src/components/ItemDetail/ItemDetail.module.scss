@import '../../styles/variables';
@import '../../styles/mediaQuery';

.container {
    @extend %container;
}

.loader {
    @extend %loader
}

.detail {
    @extend %paper;
    border-radius: 0;
    padding: $spacing1;
    display: grid;
    grid-template-columns: 1fr;
    grid-gap: $spacing1;
    @include lg {
        border-radius: $radius;
        grid-template-columns: 680px 1fr;
        grid-gap: $spacing2;
        padding: $spacing2;
    }
}

.image {
    width: 100%;
    border-radius: $radius;
}

.title {
    font-size: 20px;
    margin: 0;
    margin-bottom: 24px;
    word-break: break-all;
    @include lg {
        font-size: 24px;
        margin-bottom: $spacing1;
    }
}

.caption {
    font-size: 14px;
    color: $darkGrey;
    margin-bottom: 12px;
    display: block;
    @include lg {
        margin-bottom: $spacing1;
    }
}

.value {
    font-size: 46px;
    font-weight: 300;
    margin: 24px 0;
    @include lg {
        font-weight: 400;
        margin: $spacing2 0;
    }
    span {
        font-size: 22px;
        transform: translateY(-17px);
        display: inline-block;
        padding-left: 5px;
    }
}

.button {
    padding: 12px 10px;
    background-color: $blue;
    font-size: 15px;
    color: $white;
    border: none;
    border-radius: $radius;
    cursor: pointer;
    width: 100%;
}

.description {
    h4 {
        font-size: 24px;
        color: $black;
        margin-bottom: $spacing2;
        font-weight: 500;
        @include lg {
            font-size: 28px;
        }
    }
    p {
        font-size: 14px;
        color: $mediumGrey;
        line-height: 1.6;
        @include lg {
            font-size: 16px;
        }
    }
}