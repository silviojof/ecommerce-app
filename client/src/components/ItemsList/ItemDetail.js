import React from 'react';
import ic_shipping from '../../images/ic_shipping.png';
import ic_shipping2x from '../../images/ic_shipping@2x.png';
import Mobile1 from '../../images/mobile1.jpg';

import styles from './Items.module.scss';

const ItemDetail = ({ data }) => {
    return (
        <article className={styles.article}>
            <a href={`/items/${data.id}`} className={styles.articleGrid}>
                <div className={styles.imageContainer}>
                    <img src={Mobile1} alt="product" />
                </div>
                <div className={styles.infoContainer}>
                    <h3>
                        {`${data.price.currency} ${data.price.amount}`}
                        {
                            data.free_shipping &&
                            <img src={ic_shipping} srcSet={`${ic_shipping2x} 2x`} alt="shipping status" />
                        }
                    </h3>
                    <p>{data.title}</p>
                </div>
                <div className={styles.locationContainer}>
                    <h6>{data.location}</h6>
                </div>
            </a>
            <hr />
        </article>
    )
};

export default ItemDetail;