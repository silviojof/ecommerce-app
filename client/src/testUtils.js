import React from 'react';
import { render } from 'react-testing-library';
import { Provider } from 'react-redux';
import createSagaMiddleware from 'redux-saga';
import configureStore from 'redux-mock-store';

const sagaMiddleware = createSagaMiddleware()
const middlewares = [sagaMiddleware]

export const mockStore = configureStore(middlewares)

export const renderWithProvider = ({ store, component } = {}) => {
  return render(
    <Provider store={store}>
        {component}
    </Provider>
  )
};