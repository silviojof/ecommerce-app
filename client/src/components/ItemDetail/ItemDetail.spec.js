import React from 'react';
import {render} from 'react-testing-library';
import { renderWithProvider, mockStore } from '../../testUtils';
import { ItemDetail } from './ItemDetail';

const defaultProps = {
    location: {},
    match: {
        params: {}
    },
    fetchDetails: jest.fn(),
    details: {
        picture: '',
        condition: 'new',
        sold_quantity: 1,
        price: {
            currency: 'ARS',
            amount: 23,
            decimals: 12,
        },
        description: 'description',
    },
    clearDetail: jest.fn(),
    isLoading: false,
};

const initialStateStore = mockStore({ productList: { categories: ['first', 'second']} });

it ('matches snapshot', () => {
    const { container } = renderWithProvider({
        store: initialStateStore,
        component: (
            <ItemDetail {...defaultProps} />
        )
    });
    expect(container.firstChild).toMatchSnapshot();
});

it('isLoading', () => {
    const newProps = {
        ...defaultProps,
        isLoading: true
    };
    const {getByTestId} = render(<ItemDetail {...newProps} />);
    expect(getByTestId('loader')).toBeInTheDocument();
});

it('load Data', () => {
    const newProps = {
        ...defaultProps
    };
    const { getByTestId } = renderWithProvider({
        store: initialStateStore,
        component: (
            <ItemDetail {...newProps} />
        )
    });
    expect(getByTestId('content')).toBeInTheDocument();
});