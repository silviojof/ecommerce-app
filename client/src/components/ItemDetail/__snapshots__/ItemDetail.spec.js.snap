// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  class="container"
>
  <div
    class="breadcrumb"
  >
    <div
      class="breadcrumbItem"
    >
      <span
        class=""
      >
        first
      </span>
      <span
        class="breadcrumbArrow"
      >
        &gt;
      </span>
    </div>
    <div
      class="breadcrumbItem"
    >
      <span
        class="last"
      >
        second
      </span>
    </div>
  </div>
  <section
    class="detail"
    data-testid="content"
  >
    <img
      alt="product"
      class="image"
      src=""
    />
    <div>
      <span
        class="caption"
      >
        Nuevo - 1 vendidos
      </span>
      <h5
        class="title"
      />
      <h3
        class="value"
      >
        ARS 23
        <span>
          12
        </span>
      </h3>
      <button
        class="button"
      >
        Comprar
      </button>
    </div>
    <div
      class="description"
    >
      <h4>
        Descripci√≥n del produto
      </h4>
      <p>
        description
      </p>
    </div>
  </section>
</div>
`;
